### PR TITLE
refactor: implement COSI resource API for Talos

### PR DESCRIPTION
**Pull request by smira**
Overview: deprecate existing Talos resource API, and introduce new COSI
API.

Consequences:

* COSI API can only go via one-2-one proxy (`client.WithNode`)
* client-side API access is way easier with `state.State` wrappers
* lots of small changes on the client side to use new APIs

Signed-off-by: Andrey Smirnov <andrey.smirnov@talos-systems.com>

### CODE REVIEW

**Comment by smira** [comment]
tests are expected to fail, as some resources are not protobuf-compatible yet... waiting on a fix from @DmitriyMV

File: cmd/talosctl/cmd/talos/get.go [hunk 1]
```diff
@@ -83,210 +84,173 @@ func getResources(args []string) func(ctx context.Context, c *client.Client) err
...
 
-		if err = yaml.Unmarshal(resourceSpec, &resourceSpecRaw); err != nil {
-			continue
-		}
+		it := safe.IteratorFromList(items)
 
-		if hostname, ok := resourceSpecRaw["hostname"].(string); ok {
-			nodes = append(nodes, hostname)
-		}
+		for it.Next() {
```
Comment (frezbo) [inline cmd/talosctl/cmd/talos/get.go:241]: this looks so easy to read and no more unmarshaling :heart:

File: Dockerfile [hunk 2]
```diff
@@ -197,6 +197,8 @@ COPY ./api/cluster/cluster.proto /api/cluster/cluster.proto
 RUN protoc -I/api -I/api/vendor/ --go_out=paths=source_relative:/api --go-grpc_out=paths=source_relative:/api --go-vtproto_out=paths=source_relative:/api --go-vtproto_opt=features=marshal+unmarshal+size cluster/cluster.proto
 COPY ./api/resource/resource.proto /api/resource/resource.proto
 RUN protoc -I/api -I/api/vendor/ --go_out=paths=source_relative:/api --go-grpc_out=paths=source_relative:/api --go-vtproto_out=paths=source_relative:/api --go-vtproto_opt=features=marshal+unmarshal+size resource/resource.proto
+COPY ./api/resource/config/config.proto /api/resource/config/config.proto
```
Comment (smira) [inline Dockerfile:200]: `conifg.MachineConfig` and `config.MachineType` get custom (hand-rolled) protobuf representation, as these types are very special for generic protobuf encoder

File: api/resource/resource.proto [hunk 3]
```diff
@@ -11,6 +11,8 @@ import "google/protobuf/timestamp.proto";
 //
 // ResourceService provides user-facing API for the Talos resources.
 service ResourceService {
+  option (common.remove_deprecated_service) = "v1.5";
```
Comment (smira) [inline api/resource/resource.proto:14]: old Talos resources API is now deprecated, but fully supported until 1.5

new `talosctl` is using only COSI API

File: internal/app/apid/pkg/director/director.go [hunk 4]
```diff
@@ -62,6 +63,11 @@ func (r *Router) Director(ctx context.Context, fullMethodName string) (proxy.Mod
 
 	switch {
 	case okNodes:
+		// COSI methods do not support one-2-many proxying.
+		if strings.HasPrefix(fullMethodName, "/cosi.") {
```
Comment (smira) [inline internal/app/apid/pkg/director/director.go:67]: this is a safety belt - one2many proxy will corrupt protobufs of COSI APIs, so better to have a better explicit error

File: internal/app/machined/pkg/adapters/k8s/manifest.go [hunk 5]
```diff
@@ -68,15 +68,15 @@ func (a manifest) SetYAML(yamlBytes []byte) error {
...
 	}
 
 	return nil
 }
 
 // Objects returns list of unstructured object.
 func (a manifest) Objects() []*unstructured.Unstructured {
-	return slices.Map(a.Manifest.TypedSpec().Items, func(item map[string]interface{}) *unstructured.Unstructured {
-		return &unstructured.Unstructured{Object: item}
+	return slices.Map(a.Manifest.TypedSpec().Items, func(item k8s.SingleManifest) *unstructured.Unstructured {
```
Comment (smira) [inline internal/app/machined/pkg/adapters/k8s/manifest.go:79]: workaround `protoenc` lack of support for `[]map[string]interface{}`: wrap `map[string]interface{}` into a single-field `struct` preserving YAML marshalling

File: internal/integration/api/resources.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+// TearDownTest ...
+func (suite *ResourcesSuite) TearDownTest() {
+	if suite.ctxCancel != nil {
+		suite.ctxCancel()
+	}
+}
+
+// TestListResources tries to fetch every resource in the system.
+func (suite *ResourcesSuite) TestListResources() {
```
Comment (smira) [inline internal/integration/api/resources.go:52]: new test to fetch every resource registered

**Review by frezbo — APPROVED** [review]

File: internal/integration/api/apply-config.go [hunk 7]
```diff
@@ -462,30 +462,15 @@ func (suite *ApplyConfigSuite) TestApplyTry() {
 
 	node := nodes[0]
 
-	nodeCtx := client.WithNodes(suite.ctx, node)
+	nodeCtx := client.WithNode(suite.ctx, node)
 
 	getMachineConfig := func(ctx context.Context) (config.Provider, error) {
-		resp, err := suite.Client.Resources.Get(ctx, mc.NamespaceName, "mc", mc.V1Alpha1ID)
+		cfg, err := safe.StateGet[*mc.MachineConfig](ctx, suite.Client.COSI, resource.NewMetadata(mc.NamespaceName, mc.MachineConfigType, mc.V1Alpha1ID, resource.VersionUndefined))
```
Comment (Unix4ever) [inline internal/integration/api/apply-config.go:468]: :heart:

File: pkg/cluster/kubernetes/talos_managed.go [hunk 8]
```diff
@@ -378,52 +371,23 @@ func getManifests(ctx context.Context, cluster UpgradeProvider) ([]*unstructured
 
 	defer cluster.Close() //nolint:errcheck
 
-	listClient, err := talosclient.Resources.List(ctx, k8s.ControlPlaneNamespaceName, k8s.ManifestType)
+	md, _ := metadata.FromOutgoingContext(ctx)
+	if nodes := md["nodes"]; len(nodes) > 0 {
```
Comment (Unix4ever) [inline pkg/cluster/kubernetes/talos_managed.go:375]: Not for this PR maybe, but would be cool to use some constants for metadata fields

**Review by Unix4ever — APPROVED** [review]
:cool:

File: pkg/cluster/kubernetes/talos_managed.go (hunk 8 above)
Comment (smira) [reply pkg/cluster/kubernetes/talos_managed.go:375]: yeah, or even make this a generic loop, un-rolling `nodes` into `node` and a for loop (?)

**Pull request merged**